src/main/java/net/neoforged/neoforge/common/crafting/CraftingHelper.java

```diff
@@ -50,12 +46,15 @@ public static MapCodec<Ingredient> makeIngredientMapCodec() {
                 IngredientType::codec,
                 Ingredient.Value.MAP_CODEC)
                 .xmap(either -> either.map(ICustomIngredient::toVanilla, v -> Ingredient.fromValues(Stream.of(v))), ingredient -> {
-                    var customIngredient = convertToCompoundIngredient(ingredient).getCustomIngredient();
-                    if (customIngredient == null) {
-                        return Either.right(ingredient.getValues()[0]);
-                    } else {
-                        return Either.left(customIngredient);
+                    if (!ingredient.isCustom()) {
+                        var values = ingredient.getValues();
+                        if (values.length == 1) {
+                            return Either.right(values[0]);
+                        }
+                        // Convert vanilla ingredients with 2+ values to a CompoundIngredient. Empty ingredients are not allowed here.
+                        return Either.left(new CompoundIngredient(Stream.of(ingredient.getValues()).map(v -> Ingredient.fromValues(Stream.of(v))).toList()));
```

Comment:
What provides the guarantee that `values.length` isn't 0 here?

Reply:
It's the `.validate` call below.